refactor: tx mempool v2

mralj opened the pull request:
Continuation of https://github.com/AztecProtocol/aztec-packages/pull/15950
I have addressed most of the comments from the original [PR](https://github.com/AztecProtocol/aztec-packages/pull/15950).

Here are the ones that are "left over":

### 1. [De-dupe](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2245674781)

The reason I haven't implemented this is that I'm not convinced it will actually improve performance and will overcomplicate code. As can be seen [here](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2634974710), calling `toString` and then adding the `Fr` to Set is extremely expensive. Whereas with DB, we deal with cached memory access. 

### 2. [Bump DB version](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2355867578)
@spalladino  not sure if you still want to bump DB, or do you preferthe  current solution

### 3. [nulifier and timestamp index](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2355977978)

I'm leaving this as part of the next big refactor. 

### 4. [Insufficient fee payer balance rule](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2355955236)

There are two comments about this one, one after the other. Still, I left this code as is because I think the edge case @spalladino mentions is much less severe than the one I raised on Slack. I understand this edge case can only happen if the node was online, synced, and then went offline for some time and is now catching up. 
But in this case, they'll probably be missing non-tirval amount of txs, so this (evicting a couple of possibly valid txs) shouldn't be a problem.

Alex's idea, which I think makes more sense, is to disable pending eviction altogether during the node sync. 

**P.S.**
I have not included changes in `yarn-project/aztec/src/cli/chain_l2_config.ts` from original PR as they seem out of scope for this refactor, and I'm unsure if they are still relevant and what was the initial idea behind them. 


**P.P.S**
The issue raised on Slack referred to in point 4 was this: 

Not having the fee-payer rule as it is could make it easier to DDoS the network by attacking the mempool. This sounds scary, so I'd appreciate a sanity check. The attack is relatively simple:
1. Let's assume the attacker has a 10 balance. He'll be able to send N transactions each spending 10 balance. N could be 1k or whatever.
2. This would pass the FJ/gas balance check because FJ is checked per transaction, and each transaction is valid on its own. 
3. We don't do cumulative FJ validation because it's impossible to order transactions. As @Alex Gherghisan pointed out, ordering by nonce is not enshrined in the protocol. Any potential ordering or not accepting tx if the balance in local mempool does not cover FJ could lead only to bigger mempool divergence.
4. So all of these txs are in Schrodinger's state - both valid and invalid at the same time. I don't see what's preventing attacker(s) to fill up the mempool this way. 

If we were to include the `insufficient_fee_payer_balance rule`, then at least after the block is mined, we (soft) delete all of these TX-s from the mempool because they are invalidated by the inclusion of  one of these TXs

mralj commented:
Added a README explaining the tx pool.
Courtesy of the [amp](https://ampcode.com/threads/T-019b49f7-1a8f-7123-9acd-81c4e23d46b9). Double-checked everything, looks correct

spalladino commented on yarn-project/cli/src/config/chain_l2_config.ts:
```diff
@@ -199,7 +199,7 @@ export const stagingPublicL2ChainConfig: L2ChainConfig = {
   publicMetricsCollectorUrl: 'https://telemetry.alpha-testnet.aztec-labs.com/v1/metrics',
   publicMetricsCollectFrom: ['sequencer'],
-  maxTxPoolSize: 100_000_000, // 100MB
+  maxPendingTxCount: 100_000_000, // 100MB
```
I saw the PS on the PR description about this file, but let's update these values to something that more or less matches the old value, or at least clear the misleading comment.

PhilWindle replied:
+1

spalladino commented on yarn-project/p2p/src/client/p2p_client.ts:
```diff
@@ -828,8 +829,12 @@ export class P2PClient<T extends P2PClientType = P2PClientType.Full>
+      const blockHeader = await this.l2BlockSource.getBlockHeader(latestBlock);
+      if (!blockHeader) {
+        throw new Error(`Block header for ${latestBlock} not found`);
+      }
```
Why are we fetching the block header, if we're not using it? I'd remove this, not just for performance reasons, but also to avoid throwing in handling the event.

spalladino commented on yarn-project/foundation/src/array/sorted_array.ts:
```diff
@@ -0,0 +1,126 @@
+  // Check if we're trying to insert a duplicate
+  if (!allowDuplicates && start < arr.length && cmp(arr[start], item) === 0) {
+    return false;
+  }
```
Claude found a bug here, if you call insert with `allowDuplicates=false` on an array that already has duplicates. Granted, it's an edge case, but still worth catching.

---

## Bug Report: Incorrect duplicate check in `insertIntoSortedArray`

I found a bug in the `sortedArray` implementation at [yarn-project/foundation/src/array/sorted_array.ts:38](https://github.com/AztecProtocol/aztec-packages/blob/e9321b453592344e6c5872cc6c448ac30b724002/yarn-project/foundation/src/array/sorted_array.ts#L38).

### The Problem

The duplicate detection logic in `insertIntoSortedArray` is flawed:

```typescript
// Check if we're trying to insert a duplicate
if (!allowDuplicates && start < arr.length && cmp(arr[start], item) === 0) {
  return false;
}
```

The binary search finds an insertion point where `arr[mid] >= item`, but this doesn't guarantee it will land on an existing duplicate when multiple duplicates exist in the array.

### Example Demonstrating the Bug

```typescript
const arr = [1, 2, 2, 3];
insertIntoSortedArray(arr, 2, cmp, false);
```

The binary search might find `start = 3` (pointing to the value `3`), which means:
- The check `cmp(arr[3], 2)` evaluates to `1` (not `0`)
- The duplicate check fails
- The value `2` gets inserted: `[1, 2, 2, 2, 3]` ❌

This violates the contract when `allowDuplicates=false`.

### Recommended Fix

Check **both** the element at the insertion point AND the element before it:

```typescript
// Check if we're trying to insert a duplicate
if (!allowDuplicates) {
  // Check element before insertion point
  if (start > 0 && cmp(arr[start - 1], item) === 0) {
    return false;
  }
  // Check element at insertion point
  if (start < arr.length && cmp(arr[start], item) === 0) {
    return false;
  }
}
```

### Test Coverage Issue

The existing test "avoids inserting extra duplicates if told so" uses `[1, 2, 2, 3]` but may pass intermittently depending on which index the binary search chooses. I recommend adding more comprehensive tests with duplicates at various positions to ensure reliable detection.

mralj replied:
What time to be alive 😮

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -171,34 +193,23 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
+    });
 
-      await this.evictInvalidTxsAfterMining(txHashes, blockHeader, minedNullifiers, minedFeePayers);
+    await this.#store.transactionAsync(async () => {
```
Why on a separate transaction? What happens if another update to the tx pool gets inbetween the marking-as-mined and the eviction?

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -171,34 +193,23 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
   }
 
-  public async markMinedAsPending(txHashes: TxHash[]): Promise<void> {
+  public async markMinedAsPending(txHashes: TxHash[], blockNumber: number): Promise<void> {
```
Nit: let's rename the variable or add a comment explaining _what_ is this block number. Also, prefer using the `BlockNumber` branded type over a regular `number`.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -340,54 +328,62 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
     return this.#archivedTxLimit ? poolDbTx.then(() => this.archiveTxs(deletedTxs)) : poolDbTx;
   }
 
+  private async deleteMinedTx(txHash: `0x${string}`, minedBlockNumber: BlockNumber, permanently: boolean) {
+    await this.#minedTxHashToBlock.delete(txHash);
+    if (permanently) {
+      this.#log.trace(`Deleting mined tx ${txHash} from pool`);
+      await this.#txs.delete(txHash);
+      return;
+    }
+
+    // Soft-delete mined transactions: remove from mined set but keep in storage
+    this.#log.trace(`Soft-deleting mined tx ${txHash} from pool`);
+    await this.#deletedMinedTxHashes.set(txHash, minedBlockNumber);
+    await this.#blockToDeletedMinedTxHash.set(minedBlockNumber, txHash);
+  }
```
Shouldn't we also clear the tx from `txHashToHistoricalBlockHeaderHash`?

spalladino replied:
Oh I saw below that this index is only populated for pending txs. Let's then rename it to `pendingTxHashToHistoricalBlockHeaderHash`.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -406,37 +402,114 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
+    let historicalBlockHash = await this.#txHashToHistoricalBlockHeaderHash.getAsync(txHash.toString());
+    // Not all tx might have this index created.
+    if (!historicalBlockHash) {
+      const tx = await this.getPendingTxByHash(txHash);
+      if (!tx) {
+        throw new Error(`Tx ${txHash} not found`);
       }
```
Let's warn and return undefined instead of throwing here, I'm worried about a race condition where the tx gets deleted while we are iterating over the pending txs and constructing these infos.

mralj replied:
Done in this [commit](https://github.com/AztecProtocol/aztec-packages/pull/19014/changes/4b234a4e2b2bebf9bfec73f9c833a3ba60ead741). The reason I'm pointing it out is that I decided not to bubble up the `undefined` but filter it out. 

The only current callees of `getPendingTxInfo` are:
1.  getPendingTxInfos(): I don't see a reason not to filter out immediately the `undefined` ones here, as we are not passing any hashes or anything like that 
2. `getPendingTxsWithFeePayer(feePayers: AztecAddress[])` decision was a bit harder because we pass in `feePayers` but I don't see a clear 1:1 mapping between `feePayers` and the txInfos, so I don't see any information loss when filtering out here (instead of having `undefined` in the array)

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -569,183 +638,47 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
+      await this.#pendingTxPriorityToHash.deleteValue(getPendingTxPriority(tx), txHash);
+      this.#pendingTxs.delete(txHash);
+      await this.#historicalHeaderToTxHash.deleteValue(
+        (await tx.data.constants.anchorBlockHeader.hash()).toString(),
```
I wonder if computing the anchor block header hash so many times could have a performance impact. Let's not optimize prematurely, but we should keep an eye on it if we see the tx pool is slow-ish.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -406,37 +402,114 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
     return Promise.resolve();
   }
 
+  public clearNonEvictableTxs(): Promise<void> {
+    // Clear the non-evictable set after completing the DB updates above.
+    // This ensures pinned (non-evictable) txs are protected while we mark mined txs,
+    // but they won't remain pinned indefinitely across blocks. Note that eviction rules
+    // (including post-mining invalidation) respect the non-evictable flag while it is set.
+    this.#nonEvictableTxs.clear();
+    return Promise.resolve();
+  }
```
I know this is not from this PR, but I wonder if we may have a race condition here:

- Our RPC was slow, so we download block N at the same time we get a _proposal_ for block N+1
- The validator client processes the proposal, marking the its txs for block N+1 as non-evictable
- The p2p client processes the downloaded block N, clearing all evictable txs

Maybe the solution is to have a non-evictable flag per block?

PhilWindle replied:
I agree, this feels like a problem.

mralj replied:
I also agree, but if you think it's ok, I'd rather implement this in a separate PR - because I want to give some more thought about evicting these txs.

PhilWindle replied:
Yep, no problem.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/eviction/insufficient_fee_payer_balance_rule.ts:
```diff
@@ -0,0 +1,108 @@
+        continue;
+      }
+
+      const tx = await txPool.getTxByHash(txHash);
+      if (!tx) {
+        continue;
+      }
```
Note that we're retrieving the same tx twice, since `getPendingTxInfos` and `getPendingTxsWithFeePayer` both end up loading the entire tx into memory, then discard most of its info, and then we load it again here.

Not necessary to change now, but worth keeping in mind for a refactor where we keep in memory just the tx info we need and work from there.

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/eviction/insufficient_fee_payer_balance_rule.ts:
```diff
@@ -0,0 +1,108 @@
+      await txPool.deleteTxs(txsToEvict);
+    }
+
+    this.log.debug(`Evicted ${txsToEvict.length} invalid txs after block mined`);
```
```suggestion
    this.log.verbose(`Evicted ${txsToEvict.length} invalid txs after block mined`, { txsToEvict });
```

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/eviction/invalid_txs_after_mining_rule.ts:
```diff
@@ -0,0 +1,104 @@
+
+      const txsToEvict: TxHash[] = [];
+      const pendingTxs = await txPool.getPendingTxInfos();
+      const minedNullifiers = new Set(context.newNullifiers.map(n => n.toString()));
```
I was thinking about the performance hit of `Fr.toString`. Maybe we should use the bigint representation instead?

mralj replied:
Fixed it in I think a better way w/o any conversions (using `sortedArray`)

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/eviction/invalid_txs_after_reorg_rule.ts:
```diff
@@ -0,0 +1,87 @@
+        .filter(({ isEvictable }) => isEvictable)
+        // Multiple txs can return the same block hashes.
+        // We sort here so that if we are accessing the same leafs multiple times, we make sure they are kept in memory this way
+        // If this is randomly sorted than DB might prune already accessed leaf from the cache
```
Just in case, have you confirmed that world-state has a cache that works like that? If not, then I'd go for the deduplication, seems like the in-memory deduplication (maybe using bigint instead of string?) should be faster than multiple db accesses.

PhilWindle replied:
These queries will be going to the DB. There is no cache used for this.

mralj replied:
I'll add the dedupe, but I'm still curious if my mental model here is entirely wrong.

Cache was a clumsy word to use here; what I meant was that it won't go to disk.
The way I understand the code here is: DB is backed by LMDB, which uses `mmap`. Recently accessed pages will be in memory, so if we access the one leaf N times, it doesn't cost much, because after 1st time it should be in memory and we are "only" paying B-tree traversal. 

Even w/o sorting, this should be ok, but Alex pointed out that we might have weird access patterns here, so with sorting, we ensure we access the same leaves one after the other, hitting the in-memory page. 

I'm not convinced that dedupe will be faster than B-tree lookup, but we'd need a proper benchmark.

P.S.
I'm aware we can end up in a scenario where the memory is under intense pressure, and the OS is aggressively evicting the pages from memory.

spalladino reviewed (approved):
Great work @mralj. Left a handful of comments, take a look and feel free to merge after addressing the ones you agree with (I still think deduping is worth it btw!).

spalladino commented on yarn-project/p2p/src/mem_pools/tx_pool/README.md:
```diff
@@ -0,0 +1,241 @@
+
+## Implementations
+
+### `InMemoryTxPool` (Testing Only)
```
This reminds me that we should probably kill it, and just have a single implementation

spalladino reviewed (commented):
README looks good. I'd just add the state machine for tx states that's in the internal notion doc, which is a good resource for understanding tx lifecycle. I'd also add a line explaining the rationale for soft-deletion.

PhilWindle commented on yarn-project/end-to-end/src/spartan/mempool_limit.test.ts:
```diff
@@ -71,11 +71,11 @@ describe('mempool limiter test', () => {
     const configs = await getSequencersConfig(config);
     originalMinTxsPerBlock = configs[0]?.minTxsPerBlock;
-    await updateSequencersConfig(config, { maxTxPoolSize: 1e9, minTxsPerBlock: 0 });
+    await updateSequencersConfig(config, { maxPendingTxCount: 1e9, minTxsPerBlock: 0 });
```
This test must surely no longer work if we haven't updated the values but those values implement a completely different limit criteria now?

mralj replied:
Completely missed this one, sorry 🤦‍♂️
This is 1GB of transactions, I have changed this number to 10k transactions, which in the worst case should be 1.56GB

PhilWindle commented on yarn-project/p2p/src/mem_pools/tx_pool/aztec_kv_tx_pool.ts:
```diff
@@ -207,54 +218,24 @@ export class AztecKVTxPool extends (EventEmitter as new () => TypedEventEmitter<
 
-    await this.evictInvalidTxsAfterReorg(txHashes);
-    await this.evictLowPriorityTxs(txHashes);
+    await this.#store.transactionAsync(async () => {
```
Same comment as above. Why the new transaction?

PhilWindle commented:
Big improvement over our current implementation. There are a couple of things still to address though.

Thanks @mralj

mralj commented:
@spalladino @PhilWindle this is re-review ready :)

mralj committed:
refactor: tx mempool v2

Continuation of https://github.com/AztecProtocol/aztec-packages/pull/15950
I have addressed most of the comments from the original [PR](https://github.com/AztecProtocol/aztec-packages/pull/15950).

Here are the ones that are "left over":

### 1. [De-dupe](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2245674781)

The reason I haven't implemented this is that I'm not convinced it will actually improve performance and will overcomplicate code. As can be seen [here](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2634974710), calling `toString` and then adding the `Fr` to Set is extremely expensive. Whereas with DB, we deal with cached memory access.

### 2. [Bump DB version](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2355867578)
@spalladino  not sure if you still want to bump DB, or do you preferthe  current solution

### 3. [nulifier and timestamp index](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2355977978)

I'm leaving this as part of the next big refactor.

### 4. [Insufficient fee payer balance rule](https://github.com/AztecProtocol/aztec-packages/pull/15950#discussion_r2355955236)

There are two comments about this one, one after the other. Still, I left this code as is because I think the edge case @spalladino mentions is much less severe than the one I raised on Slack. I understand this edge case can only happen if the node was online, synced, and then went offline for some time and is now catching up.
But in this case, they'll probably be missing non-tirval amount of txs, so this (evicting a couple of possibly valid txs) shouldn't be a problem.

Alex's idea, which I think makes more sense, is to disable pending eviction altogether during the node sync.

**P.S.**
I have not included changes in `yarn-project/aztec/src/cli/chain_l2_config.ts` from original PR as they seem out of scope for this refactor, and I'm unsure if they are still relevant and what was the initial idea behind them.

**P.P.S**
The issue raised on Slack referred to in point 4 was this:

Not having the fee-payer rule as it is could make it easier to DDoS the network by attacking the mempool. This sounds scary, so I'd appreciate a sanity check. The attack is relatively simple:
1. Let's assume the attacker has a 10 balance. He'll be able to send N transactions each spending 10 balance. N could be 1k or whatever.
2. This would pass the FJ/gas balance check because FJ is checked per transaction, and each transaction is valid on its own.
3. We don't do cumulative FJ validation because it's impossible to order transactions. As @Alex Gherghisan pointed out, ordering by nonce is not enshrined in the protocol. Any potential ordering or not accepting tx if the balance in local mempool does not cover FJ could lead only to bigger mempool divergence.
4. So all of these txs are in Schrodinger's state - both valid and invalid at the same time. I don't see what's preventing attacker(s) to fill up the mempool this way.

If we were to include the `insufficient_fee_payer_balance rule`, then at least after the block is mined, we (soft) delete all of these TX-s from the mempool because they are invalidated by the inclusion of  one of these TXs

Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Phil Windle <philip.windle@gmail.com>
